rest/main.go

```diff
@@ -146,3 +168,161 @@ func serverMainPersistentConfig(fs *flag.FlagSet, flagStartupConfig *StartupConf
 
 	return false, startServer(&sc, ctx)
 }
+
+// automaticConfigUpgrade takes the config path of the current 2.x config and attempts to perform the update steps to
+// update it to a 3.x config
+// Returns the new startup config, a bool of whether to fallback to legacy config and an error
+func automaticConfigUpgrade(configPath string) (*StartupConfig, bool, error) {
+	legacyServerConfig, err := LoadServerConfig(configPath)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if legacyServerConfig.DisablePersistentConfig != nil && *legacyServerConfig.DisablePersistentConfig {
+		return nil, true, nil
+	}
+
+	startupConfig, dbConfigs, err := legacyServerConfig.ToStartupConfig()
+	if err != nil {
+		return nil, false, err
+	}
+
+	dbConfigs, err = sanitizeDbConfigs(dbConfigs)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Attempt to establish connection to server
+	cluster, err := establishCouchbaseClusterConnection(startupConfig)
+	if err != nil {
+		return nil, false, err
+	}
+
+	defer func() {
+		_ = cluster.Close()
+	}()
+
+	// Write database configs to CBS with groupID "default"
+	for _, dbConfig := range dbConfigs {
+		_, err = cluster.PutConfig(*dbConfig.Bucket, persistentConfigDefaultGroupID, base.Uint64Ptr(0), dbConfig)
+		if err != nil {
+			// If key already exists just continue
+			if base.IsAlreadyExistsError(err) {
+				continue
+			}
+			return nil, false, err
+		}
+	}
+
+	// Attempt to backup current config
+	err = backupCurrentConfigFile(configPath)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Overwrite old config with new migrated startup config
+	jsonStartupConfig, err := json.Marshal(startupConfig)
```

Comment:
pretty print JSON for new config file

```suggestion
	jsonStartupConfig, err := json.MarshalIndent(startupConfig, "", "  ")
```